services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/utils/MongoDbDeviceRegistryUtils.java

```diff
@@ -97,4 +102,43 @@ private MongoDbDeviceRegistryUtils() {
         }
         return OperationResult.empty(HttpURLConnection.HTTP_INTERNAL_ERROR);
     }
+
+    /**
+     * Checks if a provided error is {@link MongoException} with error category {@link ErrorCategory#DUPLICATE_KEY}.
+     *
+     * @param throwable The exception to check. Other Throwables but {@link MongoException} return {@code false}.
+     * @return {@code true} if exception is an {@link ErrorCategory#DUPLICATE_KEY} exception.
+     */
+    public static boolean ifDuplicateKeyError(final Throwable throwable) {
+        if (throwable instanceof MongoException) {
+            final MongoException mongoException = (MongoException) throwable;
+            return ErrorCategory.fromErrorCode(mongoException.getCode()) == ErrorCategory.DUPLICATE_KEY;
+        }
+        return false;
+    }
+
+    /**
+     * Checks if this version matches the provided version.
+     *
+     * @param <T> The type of the resource.
+     * @param resourceDto The data transfer object of the resource.
+     * @param resourceVersion The version of the resource to be matched with that of the existing one.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future will succeed if the version matches. Otherwise the future will fail with a
+     *         {@link ServiceInvocationException}.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public static <T extends BaseDto> Future<T> checkResourceVersion(final T resourceDto,
```

Comment:
It is being used only in one place inside `MongoDbBasedTenantService.getTenantNotFoundOrVersionMismatch(...)` and hence it can be removed from here. Also my suggestion is to remove this method and move the version check inside the `getTenantNotFoundOrVersionMismatch(...)` as in the `MongoDbBasedRegistrationService.checkForVersionMismatchAndFail(...)`. Meanwhile I will check how this method

Reply:
done and adapted to the method `MongoDbBasedRegistrationService.checkForVersionMismatchAndFail`.

Reply:
> Meanwhile I will check how this method `checkForVersionMismatchAndFail(...)` could be pulled up to util class so that it can be used also for tenant and credentials services.

This method has been pulled up now to `MongoDbDeviceRegistryUtils`. You could use that and remove the method definition from `MongoDbBasedRegistrationService`.

Reply:
done